### PR TITLE
feat: add gradient descent information to thermoelastic2d problem

### PR DESCRIPTION
**Pull request by gapaza**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

This pull request adds gradient information to the Thermoelastic2D problem. It also modifies the OptiStep data class to include additional optional fields for gradient information. These fields include:
- x: npt.NDArray (the current design before the gradient update)
- x_update: npt.NDArray (the update step taken by the optimizer)
- obj_values_update: npt.NDArray (how the objective values changed after the update step) 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

### Screenshots

Please attach before and after screenshots of the change if applicable.

# Checklist:

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files`
- [x] I have run `ruff check .` and `ruff format`
- [x] I have run `mypy .`
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

# Reviewer Checklist:

- [ ] The content of this PR brings value to the community. It is not too specific to a particular use case.
- [ ] The tests and checks pass (linting, formatting, type checking). For a new problem, double check the github actions workflow to ensure the problem is being tested.
- [ ] The documentation is updated.
- [ ] The code is understandable and commented. No large code blocks are left unexplained, no huge file. Can I read and understand the code easily?
- [ ] There is no merge conflict.
- [ ] The changes are not breaking the existing results (datasets, training curves, etc.). If they do, is there a good reason for it? And is the associated problem version bumped?
- [ ] For a new problem, has the dataset been generated with our slurm script so we can re-generate it if needed? (This also ensures that the problem is running on the HPC.)
- [ ] For bugfixes, it is a robust fix and not a hacky workaround.

### CODE REVIEW

File: engibench/problems/thermoelastic2d/model/fea_model.py [hunk 1]
```diff
@@ -108,6 +108,54 @@ def get_filter(self, nelx: int, nely: int, rmin: float) -> tuple[coo_matrix, np.
...
+        Returns:
+            bool: True if the optimization has converged, False otherwise. Convergence is defined as either:
+                - The change in design variables is below a predefined threshold and a minimum number of iterations have been completed.
+                - The maximum number of iterations has been reached.
+        """
+        if iterr >= self.max_iter:
+            return True
+        return change < UPDATE_THRESHOLD and iterr >= MIN_ITERATIONS
+
+    def record_step(
```
Comment (g-braeunlich) [inline engibench/problems/thermoelastic2d/model/fea_model.py]: PLR0913 starts getting annoying.
I would say: lets remove that check.
Could you do
```diff
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -132,6 +132,7 @@ ignore = [
   "EM102",    # f-string-in-exception
   "E741",     # ambiguous-variable-name
   "FIX002",   # flake8-fixme (flake8-todos is enough)
+  "PLR0913",  # too-many-arguments
   "PTH",      # flake8-use-pathlib
```
and remove the corresponding `noqa` throughout the codebase?

**Review by g-braeunlich — CHANGES_REQUESTED** [review]

File: engibench/core.py [hunk 2]
```diff
@@ -24,6 +24,11 @@ class OptiStep:
     obj_values: npt.NDArray
     step: int
 
+    # Additional Gradient Fields
+    x: npt.NDArray | None = None  # the current design before the gradient update
+    x_update: npt.NDArray | None = None  # the gradient update step taken by the optimizer
+    obj_values_update: npt.NDArray | None = None  # how the objective values change after the update step
```
Comment (gapaza) [inline engibench/core.py]: Fixed, field comments are now properly placed

File: engibench/problems/thermoelastic2d/model/fea_model.py (hunk 1 above)
Comment (gapaza) [reply engibench/problems/thermoelastic2d/model/fea_model.py]: I just ended up fixing the PLR0913 error in my function. But I see there are 14 instances of this flag being used in other problems.

Comment (gapaza) [reply engibench/problems/thermoelastic2d/model/fea_model.py]: Let me know if this is fine and I will resolve this issue

Comment (g-braeunlich) [reply engibench/problems/thermoelastic2d/model/fea_model.py]: If this is fine for your problem, then lets keep it for now

**Review by g-braeunlich — APPROVED** [review]
